### PR TITLE
py-statsmodels: add 0.14.0

### PR DESCRIPTION
**Pull request by manuelakuhn**
https://github.com/statsmodels/statsmodels/tree/v0.14.0

Installation with `install --test=root` fails with the error
```
ImportError while loading conftest '$spack-stage/spack-stage-py-statsmodels-0.14.0-olftbomdzkwcpvpm66yt4fp7vj5arooi/spack-src/build/lib.linux-x86_64-cpython-310/statsmodels/conftest.py'.
_pytest.pathlib.ImportPathMismatchError: ('statsmodels.conftest', '$spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-statsmodels-0.14.0-olftbomdzkwcpvpm66yt4fp7vj5arooi/lib/python3.10/site-packages/statsmodels/conftest.py', PosixPath('$spack-stage/spack-stage-py-statsmodels-0.14.0-olftbomdzkwcpvpm66yt4fp7vj5arooi/spack-src/build/lib.linux-x86_64-cpython-310/statsmodels/conftest.py'))
```

but installation without `--test=root` and running `spack test run py-statsmodels` afterwards, works just fine.

Any idea how to fix this?

### CODE REVIEW

**Review by tldahlgren — COMMENTED** [review]
Confirmed the version sha256.

**Comment by tldahlgren** [comment]
> https://github.com/statsmodels/statsmodels/tree/v0.14.0
> 
> Installation with `install --test=root` fails with the error
> 
> ```
> ImportError while loading conftest '$spack-stage/spack-stage-py-statsmodels-0.14.0-olftbomdzkwcpvpm66yt4fp7vj5arooi/spack-src/build/lib.linux-x86_64-cpython-310/statsmodels/conftest.py'.
> _pytest.pathlib.ImportPathMismatchError: ('statsmodels.conftest', '$spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-statsmodels-0.14.0-olftbomdzkwcpvpm66yt4fp7vj5arooi/lib/python3.10/site-packages/statsmodels/conftest.py', PosixPath('$spack-stage/spack-stage-py-statsmodels-0.14.0-olftbomdzkwcpvpm66yt4fp7vj5arooi/spack-src/build/lib.linux-x86_64-cpython-310/statsmodels/conftest.py'))
> ```
> 
> but installation without `--test=root` and running `spack test run py-statsmodels` afterwards, works just fine.
> 
> Any idea how to fix this?

IME it means there's something wrong with the default modules list so it has to be overridden (see https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#import-tests).

**Review by adamjstewart — DISMISSED** [review]

**Comment by adamjstewart** [comment]
Yes, `conftest.py` is a configuration file for pytest that is only used for testing. You can safely add it to `skip_modules`.

**Comment by tldahlgren** [comment]
> Yes, `conftest.py` is a configuration file for pytest that is only used for testing. You can safely add it to `skip_modules`.

Oops. Forgot about `skip_modules`, should've scrolled down further in the docs.

**Comment by manuelakuhn** [comment]
I tried to add it to `skip_modules` but that does not change anything (same error). It seems that that part is executed before the actual tests start (at least there is no `Testing package ...` or `Running install-time tests` in the output). What I also find strange is that `--test=root` fails but `spack test run` works.

Here the full output after the the installation again:
```
Successfully installed statsmodels-0.14.0
Removed build tracker: '/tmp/pip-build-tracker-3gansai2'
==> [2023-08-01-23:47:20.027276] '$spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-pytest-7.3.2-6xhokhynx55rvimjtltkclkedcqegftu/bin/pytest' 'statsmodels'
Copy on Write disabled
ImportError while loading conftest '$spack-stage/spack-stage-py-statsmodels-0.14.0-lzgw3xxm5txur56u5zctl4xsqij27yag/spack-src/build/lib.linux-x86_64-cpython-310/statsmodels/conftest.py'.
_pytest.pathlib.ImportPathMismatchError: ('statsmodels.conftest', '$spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-statsmodels-0.14.0-lzgw3xxm5txur56u5zctl4xsqij27yag/lib/python3.10/site-packages/statsmodels/conftest.py', PosixPath('$spack-stage/spack-stage-py-statsmodels-0.14.0-lzgw3xxm5txur56u5zctl4xsqij27yag/spack-src/build/lib.linux-x86_64-cpython-310/statsmodels/conftest.py'))
```

Ok, after some more testing: The problem is not the import tests but the additional pytest. If I comment out the `def build_test` function (the whole block), then there are no errors.

**Comment by adamjstewart** [comment]
You could try removing the directory change. https://www.statsmodels.org/stable/dev/test_notes.html runs its tests directly in the root directory.

**Comment by alecbcs** [comment]
ping @manuelakuhn

**Comment by manuelakuhn** [comment]
Rebased it and solved conflicts.

Do we still need the pinning for python here? I was not sure.

File: var/spack/repos/builtin/packages/py-statsmodels/package.py [hunk 1]
```diff
@@ -23,36 +26,50 @@ class PyStatsmodels(PythonPackage):
...
-    depends_on("py-cython@0.29.14:", type="build", when="@0.12.0:")
-    depends_on("py-cython@0.29.22:", type="build", when="@0.13.0:")
-    depends_on("py-cython@0.29.32:", type="build", when="@0.13.5:")
+    # pyproject.toml
+    depends_on("py-cython@0.29.26:2", when="@0.14:", type="build")
+    depends_on("py-cython@0.29.32:2", when="@0.13.5:0.13", type="build")
+    depends_on("py-cython@0.29.22:2", when="@0.13:", type="build")
+    depends_on("py-cython@0.29.14:2", when="@0.12:", type="build")
+    depends_on("py-cython@0.29:2", type="build")
+    depends_on("py-setuptools-scm+toml@7", when="@0.13.3:", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-statsmodels/package.py]:
```suggestion
    depends_on("py-setuptools-scm+toml@7.0", when="@0.13.3:", type="build")
```

File: var/spack/repos/builtin/packages/py-statsmodels/package.py [hunk 2]
```diff
@@ -23,36 +26,50 @@ class PyStatsmodels(PythonPackage):
     version("0.10.2", sha256="9cd2194c6642a8754e85f9a6e6912cdf996bebf6ff715d3cc67f65dadfd37cc9")
     version("0.10.1", sha256="320659a80f916c2edf9dfbe83512d9004bb562b72eedb7d9374562038697fa10")
 
-    depends_on("python@2.7:2.8,3.4:", when="@0.10.1:", type=("build", "link", "run"))
-    depends_on("python@3.6:", when="@0.12.1:", type=("build", "link", "run"))
+    # limits according to wheels available on PyPI
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-statsmodels/package.py]: Not sure if we want this or not, up to you

File: var/spack/repos/builtin/packages/py-statsmodels/package.py [hunk 3]
```diff
@@ -23,36 +26,50 @@ class PyStatsmodels(PythonPackage):
...
-    depends_on("py-scipy@0.18:", type=("build", "run"), when="@0.10.1:")
-    depends_on("py-scipy@1.2:", type=("build", "run"), when="@0.12.0:")
-    depends_on("py-scipy@1.3:", type=("build", "run"), when="@0.13.0:")
-    depends_on("py-packaging@21.3:", type=("build", "run"), when="@0.13.2:")
+    # requirements.txt
+    depends_on("py-numpy@1.18:", when="@0.14:", type=("build", "link", "run"))
+    depends_on("py-numpy@1.17:", when="@0.13:", type=("build", "link", "run"))
+    depends_on("py-numpy@1.15:", when="@0.12.1:", type=("build", "link", "run"))
+    depends_on("py-numpy@1.11:", when="@0.10.1:", type=("build", "link", "run"))
+    depends_on("py-scipy@1.4:1.9.1,1.9.3:", when="@0.13.5:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-statsmodels/package.py]: I usually do:
```suggestion
    depends_on("py-scipy@1.4:", when="@0.13.5:", type=("build", "run"))
    conflicts("^py-scipy@1.9.2")
```
for this kind of thing. If it's broken in newer releases, it's prob also broken in older releases too.

File: var/spack/repos/builtin/packages/py-statsmodels/package.py (hunk 2 above)
Comment (manuelakuhn) [reply var/spack/repos/builtin/packages/py-statsmodels/package.py]: I tested for example `py-statsmodels@0.13.2` builds fine with `python@3.11` contradicting these limitations, so I removed them.

File: var/spack/repos/builtin/packages/py-statsmodels/package.py [hunk 4]
```diff
@@ -23,36 +26,44 @@ class PyStatsmodels(PythonPackage):
...
-    depends_on("python@2.7:2.8,3.4:", when="@0.10.1:", type=("build", "link", "run"))
-    depends_on("python@3.6:", when="@0.12.1:", type=("build", "link", "run"))
+    depends_on("python@3.8:", when="@0.14:", type=("build", "link", "run"))
+    depends_on("python", type=("build", "link", "run"))
 
-    # according to https://www.statsmodels.org/dev/install.html earlier versions might work.
-    depends_on("py-setuptools", type="build")
-    depends_on("py-setuptools@59.2.0:", type="build", when="@0.13.5:")
+    # according to https://www.statsmodels.org/dev/install.html earlier versions
+    # might work.
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-statsmodels/package.py]: This comment is outdated and can prob be removed

**Review by adamjstewart — DISMISSED** [review]
Feel free to merge

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**